src/libraries/Microsoft.PowerFx.Core/Types/DType.cs

```diff
@@ -1193,6 +1199,10 @@ private bool TryGetTypeCore(DName name, out DType type)
                 case DKind.OptionSet:
                 case DKind.View:
                     return TypeTree.TryGetValue(name, out type);
+
+                case DKind.LazyRecord:
+                case DKind.LazyTable:
+                    return LazyTypeProvider.TryGetFieldType(name, out type);
```

Comment:
This is the core piece that enables this feature. All field accesses go through this function, and are delegated to the LazyTypeProvider.